torch/_dynamo/trace_rules.py

```diff
@@ -35,40 +44,155 @@
+"""
+
+
 @functools.lru_cache(None)
 def get_torch_obj_rule_map():
     d = dict()
     for k, v in torch_name_rule_map.items():
-        obj = load_object(k)
-        assert obj not in d
-        d[obj] = v
+        try:
+            obj = load_object(k)
+            d[obj] = v
+        except (AttributeError, ModuleNotFoundError):
+            pass
     return d
 
 
+def _load_obj_from_str(fully_qualified_name):
+    module, obj_name = fully_qualified_name.rsplit(".", maxsplit=1)
+    return getattr(importlib.import_module(module), obj_name)
+
+
+"""
+Load string represented torch objects.
+"""
+
+
 def load_object(name):
-    mod_name, obj_name = name.rsplit(".", 1)
-    mod = importlib.import_module(mod_name)
-    obj = getattr(mod, obj_name)
-    return obj
+    x = name.split("#")
+    if len(x) == 2:
+        obj = _load_obj_from_str(x[0])
+        val = getattr(obj, x[1])
+    else:
+        assert len(x) == 1, f"Invalid obj name {name}"
+        val = _load_obj_from_str(x[0])
+    if hasattr(val, "__wrapped__"):
+        val = val.__wrapped__
+    return val
+
+
+"""
+Get all torch.Tensor methods which are allowed to be in graph functions.
+"""
+
+
+@functools.lru_cache(None)
+def get_tensor_method():
+    s = set()
+    for name in dir(torch.Tensor):
+        method = getattr(torch.Tensor, name)
+        if isinstance(
+            method, (types.MethodDescriptorType, types.WrapperDescriptorType)
+        ):
+            s.add(method)
+    return frozenset(s)
+
+
+"""
+Return if a torch object is in graph function during Dynamo tracing.
+Note: This is a temporary function, we will have the dumped list of all in graph functions later.
+"""
+
+
+def is_in_graph_function(obj):
+    if obj in get_tensor_method() or isinstance(
+        obj,
+        (torch._ops.OpOverloadPacket, torch._ops.OpOverload),
+    ):
+        return True
+    if isinstance(
+        obj,
+        (
+            types.FunctionType,
+            types.MethodType,
+            types.BuiltinFunctionType,
+            types.MethodDescriptorType,
+            types.WrapperDescriptorType,
+        ),
+    ):
+        return is_allowed(obj)
+    else:
+        return False
+
+
+"""
+Main entry point for looking up the trace rule (the Dynamo variable) for a given object.
+E.g, the lookup result of `torch.amp.autocast_mode.autocast` is `TorchCtxManagerClassVariable`.
+"""
 
 
 def lookup(obj):
     if not hashable(obj):
         return None
+    # Custom allow/disallow in graph takes precedence over the `torch_name_rule_map`.
+    if id(obj) in _disallowed_function_ids:
+        return None
+    if is_user_defined_allowed(obj):
+        return TorchInGraphFunctionVariable
+    if hasattr(obj, "__wrapped__"):
+        # TODO: Weird case, should not unwrap if it's wrapped as _VariableFunctionsClass.
+        if not (
+            hasattr(obj, "__qualname__")
+            and str(obj.__qualname__).startswith("_VariableFunctionsClass")
+        ):
```

Comment:
This is used to mitigate a functorch issue/bug, I'll remove this when it's fixed.
The issue can be reproduced by https://gist.github.com/yanboliang/49e8fe8c0d7956420be28c27689cba76.
Somehow functorch wrap the function ```sin``` into ```_VariableFunctionsClass.sin```, and for this particular case we should not unwrapped to get the original ```sin```. I'll check with Richard to understand if functorch has special handling for ```_VariableFunctionsClass.xxx```.